api/extension-guides/notebook.md

```diff
@@ -0,0 +1,303 @@
+
+A `NotebookKernel` is responsible for taking a *code cell* and from it producing some output or set of outputs.
+
+A kernel can either be directly associated with a content provider by setting the `NotebookContentProvider#kernel` property, or registered globally by envoking the `vscode.registerNotebookKernel` function with ...TODO: Figure out how this function is supposed to be called :)...
+
+Samples:
+- [GitHub Issues Notebook](https://github.com/microsoft/vscode-github-issue-notebooks/blob/master/src/notebookProvider.ts): Kernel to execute queries for GitHib Issues
+- [HTTP Request Notebook](): Kernel to issue HTTP requests (TODO: Publish my extension? Where?)
+
+#### Best Practices
+
+While a kernel need only return an output, it can be desirable to set metadata on cells as it executes them to enable features like the run duration counter, execution order badge, and run status icon. For instance, a kernel's `executeCell` function might look like this:
+```ts
+async function executeCell(document: vscode.NotebookDocument, cell: vscode.NotebookCell, token: vscode.CancellationToken) {
+	try {
+		cell.metadata.runState = vscode.NotebookCellRunState.Running;
+		const start = +new Date();
+		cell.metadata.runStartTime = start;
+		cell.metadata.executionOrder = ++this.runIndex;
+		const result = await doExecuteCell(document, cell, token);
+		cell.outputs = [result];
+		cell.metadata.runState = vscode.NotebookCellRunState.Success;
+		cell.metadata.lastRunDuration = +new Date() - start;
+	} catch (e) {
+		cell.outputs = [{ outputKind: vscode.CellOutputKind.Error, ename: e.name, evalue: e.message, traceback: [e.stack] }];
+		cell.metadata.runState = vscode.NotebookCellRunState.Error;
+		cell.metadata.lastRunDuration = undefined;
+	}
+};
+```
+
+#### Output Types
+
+Outputs must be in one of three formats: Text Output, Error Output, or Rich Output. A kernel may provide multiple outputs for a single execution of a cell, in which case they will be displayed as a list.
+
+##### Text Output
+Text outputs are the most simple output format, and work much like many REPL's you may be familar with. They consist only of a `text` field, which is rendered as plain text in the cell's output element:
+```ts
+{
+	outputKind: vscode.CellOutputKind.Text,
+	text: '...'
+}
+```
+![Cell with simple text output](images/notebook/text-output.png)
+
+##### Error Output
+Error outputs are helpful for displaying runtime errors in a consistant and understandable manner. They contain `ename` and `evalue` fields for displaying the error type and message, respectively, as well as `traceback` field which takes a list of strings which get displayed like a callstack:
+```ts
+{
+	outputKind: vscode.CellOutputKind.Error,
+	ename: 'Error Name',
+	evalue: 'Error Value',
+	traceback: ['stack frame 1', 'stack frame 2', 'stack frame 3', 'stack frame 4']
+}
+```
+![Cell with error output showing error name and message, as well as a stack trace](images/notebook/error-output.png)
```

Comment:
maybe we can show errors which have ansi colors.

Reply:
noted